### PR TITLE
Performance and metrics touches on async workers

### PR DESCRIPTION
**Pull request by NelsonVides**
Right now, when using send_after, we're sending a single atom to the
process, which means, that we then don't know how to match the flush
request when it arrives, and therefore we need to act on all of them, or
explore the entire message queue to drop them when we do the selective
receive in `cancel_and_flush_timer`.

But that selective receive can potentially be expensive, when the batch
worker is really loaded and its queues are growing faster than it can
process, the code becomes really critical. Even more so when these
workers are enabled with message_queue_data=off_heap, which means that a
selective receive will go out of its heap, screwing cache locality, and
take all those heap fragments into its heap, which is all more
expensive.

To solve this, we can use start_timer instead of send_after, which will
then send a message that contains the TimerRef we get, so we know then
exactly which timeout is the active one, where we to receive any. This
way, cancel_timer doesn't need to worry about whether the timer actually
arrived or not: if it didn't, it's canceled, if it did, we won't attend
to it next time, it is simply ignored when the mailbox view arrives to
it.

One more tiny optimisation: we run cancel_timer asynchronously:
https://www.erlang.org/doc/man/erlang.html#cancel_timer-2

> The timer service that manages the timer can be co-located with
another scheduler than the scheduler that the calling process is
executing on. If so, communication with the timer service takes much
longer time than if it is located locally. If the calling process is in
critical path, and can do other things while waiting for the result of
this operation, or is not interested in the result of the operation, you
want to use option {async, true}. If using option {async, false}, the
calling process blocks until the operation has been performed.

There's other code, see commits for details.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_23](https://circleci.com/gh/esl/MongooseIM/100757) / small_tests / 0248b77f637013f1734eb3ba5f2ef279de1d7f09
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3502/100757/small_tests.23.1) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3502/100757/small_tests.23.1/small/ct_run.mongooseim@localhost.2022-01-14_17.59.49/lib.mongooseim.logs/run.2022-01-14_17.59.49/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100793) / small_tests / d6d522f4d2af7902968f8dd30a376f0ca536e77e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3502/100793/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3502/100793/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-14_18.10.42/lib.mongooseim.logs/run.2022-01-14_18.10.43/suite.log.html)

File: src/mongoose_async_pools.erl [hunk 1]
```diff
@@ -77,6 +82,8 @@ start_link(HostType, PoolId, Opts) ->
 -spec init(term()) -> {ok, {supervisor:sup_flags(), [supervisor:child_spec()]}}.
 init([HostType, PoolId, Opts]) ->
     PoolName = gen_pool_name(HostType, PoolId),
+    mongoose_metrics:ensure_metric(HostType, [?MODULE, PoolId, timed_flushes], counter),
```
Comment (arcusfelis) [inline src/mongoose_async_pools.erl:85]: can we have a test that it gets incremented?

**Review by arcusfelis — APPROVED** [review]
looks very good, need a test for metrics

File: src/mongoose_async_pools.erl (hunk 1 above)
Comment (NelsonVides) [reply src/mongoose_async_pools.erl:85]: Metric test implemented

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/100985) / small_tests / 0725fc242a1896c10e73e4cd40398be9d3b703a3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3502/100985/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3502/100985/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-17_17.36.37/lib.mongooseim.logs/run.2022-01-17_17.36.37/suite.log.html)

File: big_tests/tests/mam_SUITE.erl [hunk 2]
```diff
@@ -3074,6 +3084,24 @@ metric_incremented_on_archive_request(ConfigIn) ->
...
+
+get_mongoose_async_metrics() ->
+    HostType = domain_helper:host_type(mim),
+    HostTypePrefix = domain_helper:make_metrics_prefix(HostType),
+    #{batch_flushes => get_mongoose_async_metrics(HostTypePrefix, batch_flushes),
+      timed_flushes => get_mongoose_async_metrics(HostTypePrefix, timed_flushes)}.
+
+get_mongoose_async_metrics(HostTypePrefix, MetricName) ->
+    Metric = [HostTypePrefix, mongoose_async_pools, pm_mam, MetricName],
+    {ok, Value} = rpc(mim(), exometer, get_value, [Metric]),
```
Comment (arcusfelis) [inline big_tests/tests/mam_SUITE.erl:3101]: use `mongoose_metrics:get_metric_value/1`. Just in case we move away from exometer :)

Comment (NelsonVides) [reply big_tests/tests/mam_SUITE.erl]: Done 👍🏽

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/101080) / small_tests / 4572526dafec81dfd893ffb77967721956188655
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3502/101080/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3502/101080/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-01-17_18.05.42/lib.mongooseim.logs/run.2022-01-17_18.05.42/suite.log.html)

**Pull request merged**